__tests__/OmnichannelChatSDK.spec.ts

```diff
@@ -476,5 +994,80 @@ describe('Omnichannel Chat SDK', () => {
+            chatSDK.getChatConfig = jest.fn();
+            chatSDK.getChatToken = jest.fn();
+
+            await chatSDK.initialize();
+            await chatSDK.startChat();
+
+            jest.spyOn(chatSDK.OCClient, 'sessionInit').mockResolvedValue(Promise.resolve());
+            jest.spyOn(chatSDK.IC3Client, 'initialize').mockResolvedValue(Promise.resolve());
+            jest.spyOn(chatSDK.IC3Client, 'joinConversation').mockResolvedValue(Promise.resolve({
+                disconnect: () => {}
+            }));
+
+            jest.spyOn(chatSDK.OCClient, 'sessionClose').mockRejectedValue(Promise.reject());
+
+            try {
+                await chatSDK.endChat();
+            } catch (error) {
+                expect(console.error).toHaveBeenCalled();
+            }
+
+            expect(chatSDK.OCClient.sessionClose).toHaveBeenCalledTimes(1);
+        });
+
+        it('ChatSDK.endChat() with authenticatedUserToken should pass it to OCClient.sessionClose() call\'s optional parameters', async () => {
+            const chatSDKConfig = {
+                getAuthToken: async () => {
+                    return 'authenticatedUserToken'
+                }
+            };
+
+            const chatSDK = new OmnichannelChatSDK(omnichannelConfig, chatSDKConfig);
+            chatSDK.getChatConfig = jest.fn();
+
+            await chatSDK.initialize();
+
+            chatSDK.IC3Client.initialize = jest.fn();
+            chatSDK.IC3Client.joinConversation = jest.fn();
+
+            const optionaParams = {
+                authenticatedUserToken: 'authenticatedUserToken'
+            }
+
+            jest.spyOn(chatSDK.OCClient, 'getChatConfig').mockResolvedValue(Promise.resolve({
+                DataMaskingInfo: { setting: { msdyn_maskforcustomer: false } },
+                LiveWSAndLiveChatEngJoin: { PreChatSurvey: { msdyn_prechatenabled: false } },
+                LiveChatConfigAuthSettings: {}
+            }));
+
+            jest.spyOn(chatSDK.OCClient, 'getChatToken').mockResolvedValue(Promise.resolve({
+                ChatId: '',
+                Token: '',
+                RegionGtms: '{}'
+            }));
+
+            jest.spyOn(chatSDK.OCClient, 'sessionInit').mockResolvedValue(Promise.resolve());
+            jest.spyOn(chatSDK.OCClient, 'sessionClose').mockResolvedValue(Promise.resolve());
+
+            await chatSDK.startChat();
+
+            chatSDK.authenticatedUserToken = optionaParams.authenticatedUserToken;
+            chatSDK.conversation = {
+                disconnect: jest.fn()
+            };
+
+            const sessionCloseOptionalParams = {
+                authenticatedUserToken: optionaParams.authenticatedUserToken
+            }
+
+            await chatSDK.endChat();
+            expect(chatSDK.OCClient.sessionClose).toHaveBeenCalledTimes(1);
+            expect(chatSDK.OCClient.sessionClose.mock.calls[0][1]).toMatchObject(sessionCloseOptionalParams);
+        });
```

Comment:
I think the tests in OmniChannelChatSdk.test.ts is great and I'm okay if we just keep it like this so we can increase our code coverage numbers.

I have some comments though, maybe we can have a conversation about them.

There's a lot of boiler plate in each test in this file. I had some ideas regarding this.

1. What do you think of fewer tests (maybe even just one) that check the critical path of a successful conversation (i.e. startChat(), getMessages(), sendMessage(), downloadFile(), getLiveChatTranscript(),  emailLiveChatTranscript(), endChat()).

2. We could also have one that checks where various things fail (sendMessage() and downloadFile() when there's no internet connection, startChat() without a proper authenticatedUserToken, etc).

3. Wouldn't we also want a way to test registering and receiving the various events? I.e. `onNewMessage()`, `onAgentEndSession()`, `onTypingEvent()`.

Reply:
1. I agree that we should have fewer, but meaningful tests. Code coverage is not the best indicator of the test quality. We also have to maintain the tests, so having too many can have an impact in our development productivity. We'll be able to have better test quality once we'll get our e2e test harness.
2. Agree. These tests would be meaningful. It should be part of e2e tests.
3. It'd be a good idea to test registering and receiving events. We'll have to find a way on how to test them.